### DISCUSSION TITLE
1.8.0 Pre-Release + plugin API breaking change backwards compatibility discussion

### DISCUSSION THREAD

**Discussion by jjw24** [Flow-Launcher/Flow.Launcher#535] [Announcements]
Hi all,

1.8.0 is our upcoming release. You can test it by clicking [here](https://ci.appveyor.com/api/buildjobs/bh1p8970v3he3krp/artifacts/Output%2FPackages%2FFlow-Launcher-v1.7.2.exe) to download the installer or you can check the release notes [here](https://github.com/Flow-Launcher/Flow.Launcher/pull/363).

Make sure your Flow is not in Portable Mode otherwise your user data will be lost. Please note until the official release, the version of this installation will still be marked as 1.7.2, this is so when it's officially released or newer versions are out you can still update.

Let us know if you experience any issues, and please make sure you let us know that you are running the pre-release version.

In the future, if you would like to download and test a specific build from the PR, simply go to the PR you would like and find the latest commit with a green tick and click 'Details':
![image](https://user-images.githubusercontent.com/26427004/123897985-c7d01200-d9a7-11eb-959f-4a95c4105a97.png)

This will take you to our build page. Click on 'Artifacts' then select 'Output\Packages\Flow-Launcher-v1.7.2.exe':
![image](https://user-images.githubusercontent.com/26427004/123898100-0cf44400-d9a8-11eb-80d0-3d90be0e89b3.png)

Enjoy.

**Comment by deefrawley** [comment]
@jjw24  I've found a reproducible crash. Should I give details here or open a new issue?

**Reply by jjw24** [reply]
Yeah, if it's confirmed crash bug, chuck it in issues, title it 'dev branch:' + the description so we know it's only from dev

**Reply by deefrawley** [reply]
I've opened in my plugin repo for now as further testing tells me it is my plugin's fault somehow.

**Comment by jjw24** [comment]
Once https://github.com/Flow-Launcher/Flow.Launcher/pull/541 goes in we should be good to release 1.8.0 this weekend hopefully, what do you  peeps think?

I haven't come across any issues using it.

@Flow-Launcher/team

**Reply by pc223** [reply]
I daily drive dev branch for a couple of days now, seems to be solid 👍

**Reply by deefrawley** [reply]
The python context menu issue was the only problem I have found in the pre-release build I've been using this week. If that's fixed it's a thumbs up from me.

**Comment by JohnTheGr8** [comment]
How about we consider a more future-proof solution to the compatibility issues described in #480 ? Seems like our hands are currently tied when it comes to deprecating stuff from the plugin interfaces, and releasing 1.8.0 as-is won't make it easier to resolve in the future...

**Reply by JohnTheGr8** [reply]
Haven't given this much thought, but maybe something like different branches in the manifest repo for different versions of Flow? So that Flow `v1.18.x` loads plugin data from the `flow_1_18` branch, `v1.17.x` from the `flow_1_17` branch, and so on...

Pros that I can think of:
- plugin developers could work on/publish pre-release versions of their plugins that target pre-release versions of Flow
- no need to update PluginsManager, assuming that we properly vet what goes in the manifest repo

Cons that I can think of:
- the auto-update script in the manifest repo will need some adjustments
- might confuse some plugin developers
- the `main` branch of the manifest repo will have to remain as-is for compatibility

Again, the above is off-the-top-of-my-head, haven't given it too much thought. Thoughts?

**Reply by taooceros** [reply]
If a plugin that doesn't have a version required for Flow (no use for new api, targeting older framework), should the author update its manifest in all the branches?

**Reply by JohnTheGr8** [reply]
I think that's unavoidable with this setup, but at the same time I think it's fine for plugin developers to only focus on the latest branch, at least when it comes to plugin updates etc

Basically, branches for older versions of Flow just remain there to avoid breaking things for users that haven't updated yet. Plugin developers shouldn't be expected to deal with this unless they want to. I think it's a reasonable compromise...

**Reply by taooceros** [reply]
By the way, currently we don't actually do version update in pre-release until the final release (we haven't even till now because of the consideration of auto-update).
Though, that's not a big due, and it is possible to mark current pre-release as older than the same version of actual release (or even use 1.8.0 for pre-release and 1.8.1 for real release), and update the release number earlier.

**Reply by JohnTheGr8** [reply]
> 
> 
> By the way, currently we don't actually do version update in pre-release until the final release (we haven't even till now because of the consideration of auto-update).
> Though, that's not a big due, and it is possible to mark current pre-release as older than the same version of actual release (or even use 1.8.0 for pre-release and 1.8.1 for real release), and update the release number earlier.

the version itself is irrelevant, we can simply update [this link in PluginsManifest.cs](https://github.com/Flow-Launcher/Flow.Launcher/blob/dev/Plugins/Flow.Launcher.Plugin.PluginsManager/Models/PluginsManifest.cs#L18) as soon as we create a new branch in the manifest repository

**Reply by taooceros** [reply]
> the version itself is irrelevant, we can simply update this link in PluginsManifest.cs as soon as we create a new branch in the manifest repository

You mean to hard-code the link in PluginsManifest?

**Reply by taooceros** [reply]
> the auto-update script in the manifest repo will need some adjustments

It shouldn't be hard, and what's need to be done is simply adding a matrix including the branches that we wants them to update.

> might confuse some plugin developers

What about we still keep developers to contribute to the main (or some other) branch, and update the release and pre-release branch with the CI?

**Reply by JohnTheGr8** [reply]
> You mean to hard-code the link in PluginsManifest?

Well, it's already hard-coded, pointing to the main branch:  `..../main/plugins.json` We can instead hard-code it to a more version-specific branch name and update it whenever we introduce breaking changes

> It shouldn't be hard, and what's need to be done is simply adding a matrix including the branches that we wants them to update.

We'd probably only want to auto-update on the latest branch to avoid introducing breaking changes to the legacy branches.

> What about we still keep developers to contribute to the main (or some other) branch, and update the release and pre-release branch with the CI?

I agree plugin developers shouldn't bother with branches other than the most recent one, that's what I meant in my previous comment ("Plugin developers shouldn't be expected to deal with this unless they want to.")

**Reply by JohnTheGr8** [reply]
Here's a full example to make this clear. To keep it relevant, let's assume we had this system in place now...

The released version of Flow (master branch, v1.7.2) would point to `..../flow_1_7_x/plugins.json` in `PluginsManifest.cs`. The `flow_1_7_x` branch of the plugin manifest repo would be the default branch, and all new plugins & updates would be added to this branch while v1.8 was in development...

As soon as we introduced the breaking changes to the plugin interfaces in the `dev` branch of Flow, we would add a `flow_1_8_x` branch to the plugin manifest repo (starting as a copy of `flow_1_7_x`) and update the link in `PluginsManifest.cs`. This `flow_1_8_x` branch would act as a pre-release branch for plugins, and plugin developers would be able to develop/test/update their plugins for the upcoming 1.8.0 release of Flow before 1.8.0 is officially released.

As soon as 1.8.0 is released, we simply stop auto-updating plugins in the `flow_1_7_x` branch, we basically leave it for compatibility for those users that have not updated to the latest Flow yet. The `flow_1_8_x` is now the default branch for new plugins and updates.

In other words, branch `flow_1_7_x` would be for plugins depending on the 1.4.0 version of Flow.Launcher.Plugin, while `flow_1_8_x` would be for plugins depending on 2.0.0

@jjw24 thoughts?

**Reply by jjw24** [reply]
yes i think this would be sufficient for us right now. I have seen it done in other projects, probably be the simplest way for us right now.

only thing is we wont be able to rename main as flow 1.7.2 versions and below will be pointing to it. I can create a new branch now and make it default, and also update the code. Should we call the new branch in pluginsmanifest repo flow_1_8_x or plugin_api_v2? i like the latter because it actually reflects better what the branch is for, and a new breaking change would be plugin_api_v3.

with this method, we are aware that in order for users to use any new plugin updates they will have to upgrade to flow 1.8.0? (which i am ok because it's always good to move users to a newer version so they get the fixes and new features)

**Reply by Zeroto521** [reply]
I agree with @JohnTheGr8's. I thought FLow Launcher is a **platform**.

Flow.Launcher needs to keep a branch called `flow_1_7` (or else) after 1.8.0 released.
This branch is used to fix bugs but no features added until one day no more maintenance.

---

To handle API version conflict, the PluginsManifest side has two ways.

## For different API versions to build different plugin branches

Like this Flow-Launcher/Flow.Launcher.PluginsManifest#53 did.

Flow in `flow_1_7_x` branch install plugins in `plugin_api_v1` branch.
Flow in `flow_1_8_x` branch install plugins in `plugin_api_v2` or `main` branch.

This way for Flow is simply, just change the link to PluginsManifest.

I don't like this way, this would make the PluginsManifest dirtier and dirtier.
More branches come out with development, and that causes **hard to manage** whatever for PluginsManifest or Flow.Launcher.

## Add a new Field called `flow_api_version` (or else)

Add the field to let others know which Flow version supports this plugin.
And all plugins still in one place.

Two ways all need the plugin developer to notice flow API version.
A field could easily handle this.

This way need Flow adds some codes to handle the new field.

If `plugin.json` no this new field, we could consider this plugin is good for all API versions unless there has.

**Reply by jjw24** [reply]
Method one is easy to implement but makes maintenance eventually messy in the long term.

Any downside/cons to method two?

**Reply by jjw24** [reply]
For one, method two is not going to work for users on older versions because the new proposed field is not going to get recognized. So you need to actually implement method one first.

**Reply by jjw24** [reply]
Open to more discussions, but I am still currently in favour for using method one because plugin devs don't need to be version aware, just commit to the default branch of PluginsManifest and we take care of backwards compatibility. 

I actually don't think there will be too many branches because I would think when we go to next breaking change we repoint it back to the main branch as the default. This will also force users that are on v1 API to upgrade their flow, and this force upgrade is actually a good thing because users shouldnt even be on very old versions. There shouldn't be a reason that they want to stay on old versions and if there is we need to know about it so we can fix it.

**Reply by Zeroto521** [reply]
Maybe we could combine these two ways.

Only two branches for PluginsManifest.

One branch to keep the old plugin with no the new proposed field.
This branch for Flow v1.7.2 and the lower version.

Another branch (main) adds this new field for the future.

**Reply by JohnTheGr8** [reply]
Combining both ways seems more confusing and to require more work without much benefit at this point. 

Also, the `main` branch must continue to point to plugins that are compatible with Flow 1.7.2, this we cannot avoid...

**Comment by taooceros** [comment]
I will respond here since the reply is too much.

For current situation, method two may not be a good way, because it will make the plugins can be installed by every version of Flow by default. However, the latest Flow.Launcher.Plugin.dll may not support older version of Flow if they use some new api. Also, as @jjw24 has suggested, it won't ban current Flow (1.7.2 or older) from receiving plugins update.

It would be quite possible for a developer to download the latest dll, but using some basic feature at first, which works for older version of Flow, but one update brings the use of newer feature that is not supported by older Flow, and users update plugin without updating Flow, and the plugin become unusable. It would be hard for them to check which version adds some new feature and add the compatibility information there. 

Furthermore, this means that one plugin may need to have multiple version listed in the manifest that supports different version of Flow if they want, which will make the manifest even messy.


By the way, I think we should only use reply when it is a simple reply. When a new idea or a new proposal need to be discussed, it would be better to open another field to discuss. The vote feature can also be used in this way.

**Comment by jjw24** [comment]
So in summary we have considered the following two methods for maintaining API's backwards compatibility with older versions of Flow:

**Method one:**
Branch system, in PluginsManifest repo, a separate branch for the new upcoming API v2 release and will be the default. This will help older versions of Flow pointing to the main branch continue to work because plugins can still be downloaded from there. New plugin updates however will not be available for Flow users on versions before 1.8.0 because they will be pointing to the main branch and new plugin updates will be committed to the new default branch for API v2.

Users upgraded to 1.8.0 will receive new plugin updates including plugins using API v2 as their flow will be pointing to the new API v2 branch in PluginsManifest.

We will try to maintain no more than these two branches (main + new v2 api branch) in PluginsManifest to help with maintainability, and with another breaking change API update i.e. API v3 we will point the url in Flow back to the main branch. This will eventually break users that are still on 1.7.2 and older versions of Flow, but it is ok because as soon as they update they their plugins will work again. Users should let us know if there is a reason that they have been using 1.7.2 or older so we can fix it.

**Method two**   
Use just one branch (main) in PluginsManifest, add a field that indicates the version of Flow the plugin supports. Add code in Flow to recognize this field and only update the plugin if the current flow version is supported. This way no additional branches and all plugins in one place. 

Plugin developer will need to be aware what flow version their plugins support.

**Conclusion**
Method two can not be implemented because there is no code in flow 1.7.2 and older to recognise the new field, so backwards compatibility with the new API can not be achieved. Users still on 1.7.2 and older when updating to a plugin that is using the new v2 API will break the plugin and make it unusable. Additional downsides see above reply and the above conversation thread.

Method one will allow users on 1.7.2 and older to be able to download all plugins that are added prior to plugin API v2 release, but will not receive new updates or able to install newly added plugins after api v2 release. This can be addressed as soon as they update flow to the latest. Branches for implementing this method in PluginsManifest repo should be kept to two, main + new api v2 branch, and repoint to main branch once a new API major version (breaking) is released, which would not be in the immediate short term. Additionally plugin developers do not need to care what api version their new update supports.

It seems method one is quicker to implement and as long as we don't allow too many branches, it should be a good approach. Do we need to discuss this further, are there any concerns not addressed, or are we good to implement it? @Flow-Launcher/team

**Comment by taooceros** [comment]
Besides this, what are something we still need to work on for 1.8.0 release?

**Reply by jjw24** [reply]
do we need to resolve @pc223 's issue https://github.com/Flow-Launcher/Flow.Launcher/discussions/553

**Reply by taooceros** [reply]
> do we need to resolve @pc223 's issue https://github.com/Flow-Launcher/Flow.Launcher/discussions/553

I don't think so. 

It will take a while to resolve it because I have no idea how to do so...

I think it's a quite conner case (seldom people need to have 5000 results at startup), and workaround is exist so it shouldn't affect too much.

**Reply by pc223** [reply]
@taooceros Do you think PR #554 ready to add to 1.8.0 as a temporary fix? I use it for couple of days now, seems pretty good

**Reply by taooceros** [reply]
Let me take some check on that PR. If the delay only happens with debugger, we can add it as a temporary fix.

**Reply by taooceros** [reply]
@pc223 I add a warm up for #554. Would you please take a try? Though, this freeze is outside my knowledge scope, so I can only experiment some potential 😂

**Reply by pc223** [reply]
> @pc223 I add a warm up for #554. Would you please take a try? Though, this freeze is outside my knowledge scope, so I can only experiment some potential 😂

Sure, I'll try it out later 😄
